be/src/exec/pipeline/bucket_process_operator.cpp

```diff
@@ -0,0 +1,157 @@
+#include "exec/pipeline/bucket_process_operator.h"
+
+#include "exec/pipeline/operator.h"
+#include "exec/pipeline/pipeline_fwd.h"
+#include "runtime/runtime_state.h"
+#include "util/runtime_profile.h"
+
+namespace starrocks::pipeline {
+
+Status BucketProcessContext::reset_operator_state(RuntimeState* state) {
+    RETURN_IF_ERROR(source->reset_state(state, {}));
+    RETURN_IF_ERROR(sink->reset_state(state, {}));
+    return Status::OK();
+}
+
+Status BucketProcessSinkOperator::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(Operator::prepare(state));
+    RETURN_IF_ERROR(_ctx->sink->prepare(state));
+    return Status::OK();
+}
+
+void BucketProcessSinkOperator::close(RuntimeState* state) {
+    _ctx->sink->close(state);
+}
+
+bool BucketProcessSinkOperator::need_input() const {
+    if (_ctx->current_bucket_sink_finished) {
+        return false;
+    }
+    return _ctx->sink->need_input();
+}
+
+bool BucketProcessSinkOperator::is_finished() const {
+    return _ctx->all_input_finishing && _ctx->sink->is_finished();
+}
+
+Status BucketProcessSinkOperator::set_finishing(RuntimeState* state) {
+    _ctx->all_input_finishing = true;
+    bool token = _ctx->token;
+    if (!token && _ctx->token.compare_exchange_strong(token, true)) {
+        RETURN_IF_ERROR(_ctx->sink->set_finishing(state));
+        _ctx->current_bucket_sink_finished = true;
```

Comment:
should set even if error?

Reply:
we don't have to set it. we keep _ctx->all_input_finishing = true; in set_finishing is enough